independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java

```diff
@@ -63,226 +63,222 @@ Collection<Resource> generate(DotName scope) {
-        implementForEach(contextInstances, idToField, lockField.getFieldDescriptor());
+        implementComputeIfAbsent(contextInstances, beans, idToFields);
+        implementGetIfPresent(contextInstances, beans, idToFields);
+        implementRemove(contextInstances, beans, idToFields);
+        implementGetAllPresent(contextInstances, idToFields);
+        implementRemoveEach(contextInstances, idToFields);
+
+        // These methods are needed to significantly reduce the size of the stack map table for getAllPresent() and removeEach()
+        implementLockAll(contextInstances, idToFields);
+        implementUnlockAll(contextInstances, idToFields);
+
         contextInstances.close();
 
         return classOutput.getResources();
     }
 
-    private void implementGetAllPresent(ClassCreator contextInstances, Map<String, FieldDescriptor> idToField,
-            FieldDescriptor lockField) {
+    private void implementGetAllPresent(ClassCreator contextInstances, Map<String, InstanceAndLock> idToFields) {
         MethodCreator getAllPresent = contextInstances.getMethodCreator("getAllPresent", Set.class)
                 .setModifiers(ACC_PUBLIC);
-        // lock.lock();
-        // try {
-        //    Set<ContextInstanceHandle<?>> ret = new HashSet<>();
-        //    ContextInstanceHandle<?> copy = this.1;
-        //    if (copy != null) {
-        //       ret.add(copy);
-        //    }
-        //    return ret;
-        // } catch(Throwable t) {
-        //    lock.unlock();
-        //    throw t;
+        // this.lockAll();
+        // Set<ContextInstanceHandle<?>> ret = new HashSet<>();
+        // ContextInstanceHandle<?> copy1 = this.1;
+        // if (copy1 != null) {
+        //    ret.add(copy1);
+        // }
+        // this.unlockAll();
+        // return ret;
         // }
-        ResultHandle lock = getAllPresent.readInstanceField(lockField, getAllPresent.getThis());
-        getAllPresent.invokeInterfaceMethod(MethodDescriptors.LOCK_LOCK, lock);
-        TryBlock tryBlock = getAllPresent.tryBlock();
-        ResultHandle ret = tryBlock.newInstance(MethodDescriptor.ofConstructor(HashSet.class));
-        for (FieldDescriptor field : idToField.values()) {
-            ResultHandle copy = tryBlock.readInstanceField(field, tryBlock.getThis());
-            tryBlock.ifNotNull(copy).trueBranch().invokeInterfaceMethod(MethodDescriptors.SET_ADD, ret, copy);
+        getAllPresent.invokeVirtualMethod(MethodDescriptor.ofMethod(contextInstances.getClassName(), "lockAll", void.class),
+                getAllPresent.getThis());
+        ResultHandle ret = getAllPresent.newInstance(MethodDescriptor.ofConstructor(HashSet.class));
+        for (InstanceAndLock fields : idToFields.values()) {
+            ResultHandle copy = getAllPresent.readInstanceField(fields.instance, getAllPresent.getThis());
+            getAllPresent.ifNotNull(copy).trueBranch().invokeInterfaceMethod(MethodDescriptors.SET_ADD, ret, copy);
```

Comment:
This could in theory throw `OutOfMemoryError` or something similarly unlikely, so maybe we should wrap this all in a `try/catch` and `unlockAll` in the `catch` block?

Reply:
Well, if there's an `OutOfMemoryError` I don't think it would help a lot...

Reply:
We could alternatively do what `removeEach` does: under the lock, only copy the fields into local variables, and do everything else after unlocking.

Reply:
Done.